### PR TITLE
add sys admin screen for creating realms

### PR DESCRIPTION
**Pull request by mikehelmick**
## Proposed Changes

* System admins can create new realms (very basic)
* clear user from cache when realms change - otherwise you can't select the new realm until record expires.

**Release Note**

```release-note
System admin page to create realms from the UI instead of command line tools.
```

### CODE REVIEW

File: cmd/server/assets/admin/new.html [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  {{template "head" .}}
+</head>
+
+<body>
+  {{template "navbar" .}}
+
+  <main role="main" class="container">
+    {{template "flash" .}}
+
+    <h1>New Realm</h1>
```
Comment (sethvargo) [inline cmd/server/assets/admin/new.html:17]:
```suggestion
    <h1>New realm</h1>
```

File: cmd/server/assets/admin/new.html [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+          <div class="form-group row">
+            <label for="name" class="col-sm-3">Name:</label>
+            <div class="col-sm-9">
+              <input type="text" id="name" name="name" class="form-control{{if $realm.ErrorsFor "name"}} is-invalid{{end}}" value="{{$realm.Name}}">
+              {{if $realm.ErrorsFor "name"}}
+              <div class="invalid-feedback">
+                {{joinStrings ($realm.ErrorsFor "name") ", "}}
+              </div>
+              {{end}}
```
Comment (sethvargo) [inline cmd/server/assets/admin/new.html:36]: Should we add the help text beneath?

File: cmd/server/assets/admin/new.html [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            </div>
+          </div>
+
+          <div class="form-group row">
+            <label for="name" class="col-sm-3">Region Code:</label>
```
Comment (sethvargo) [inline cmd/server/assets/admin/new.html:41]:
```suggestion
            <label for="name" class="col-sm-3">Region code:</label>
```

File: cmd/server/assets/admin/new.html [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+              {{if $realm.ErrorsFor "regionCode"}}
+              <div class="invalid-feedback">
+                {{joinStrings ($realm.ErrorsFor "regionCode") ", "}}
+              </div>
+              {{end}}
+            </div>
+          </div>
+
+          <div class="form-group row">
+            <label for="type" class="col-sm-3">Certificate Signing Keys:</label>
```
Comment (sethvargo) [inline cmd/server/assets/admin/new.html:53]:
```suggestion
            <label for="type" class="col-sm-3">Certificate signing keys:</label>
```

File: cmd/server/assets/admin/new.html [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            <div class="col-sm-9">
+              <select class="form-control{{if $realm.ErrorsFor "UseRealmCertificateKey"}} is-invalid{{end}}" name="useRealmCertificateKey" id="useRealmCertificateKey">
+                <option value="true" {{if $realm.UseRealmCertificateKey}}selected{{end}}>Create realm specific signing key</option>
+                <option value="false" {{if not $realm.UseRealmCertificateKey}}selected{{end}}>Use system signing key</option>
+              </select>
+              {{if $realm.ErrorsFor "UseRealmCertificateKey"}}
+              <div class="invalid-feedback">
+                {{joinStrings ($realm.ErrorsFor "UseRealmCertificateKey") ", "}}
+              </div>
+              {{end}}
```
Comment (sethvargo) [inline cmd/server/assets/admin/new.html:63]: Definitely want help text for this one

File: cmd/server/assets/admin/realms.html [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+       These are the currently available realms. You can also <a href="/admin/realms/create">create a new realm</a>.
+    </p>
+    
+    {{if .realms}}
+    <div class="table-responsive">
+      <table class="table table-bordered table-striped">
+        <thead>
+          <tr>
+            <th scope="col">Name</th>
+            <th scope="col" width="25">Region Code</th>
```
Comment (sethvargo) [inline cmd/server/assets/admin/realms.html:26]:
```suggestion
            <th scope="col" width="25">Region code</th>
```

File: cmd/server/assets/admin/realms.html [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            <th scope="col" width="50">Signing Key</th>
```
Comment (sethvargo) [inline cmd/server/assets/admin/realms.html:27]:
```suggestion
            <th scope="col" width="50">Signing key</th>
```

File: cmd/server/assets/header.html [hunk 8]
```diff
@@ -189,6 +189,12 @@ <h6 class="dropdown-header">Manage realm</h6>
               {{end}}
               {{end}}
 
+              {{if .currentUser.Admin}}
+              <h6 class="dropdown-header">System Admin</h6>
```
Comment (sethvargo) [inline cmd/server/assets/header.html:193]:
```suggestion
              <h6 class="dropdown-header">System</h6>
```

File: pkg/controller/admin/admin.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+type Controller struct {
+	config *config.ServerConfig
+	cacher cache.Cacher
+	db     *database.Database
+	h      *render.Renderer
+	logger *zap.SugaredLogger
+}
+
+func New(ctx context.Context, config *config.ServerConfig, cacher cache.Cacher, db *database.Database, h *render.Renderer) *Controller {
+	logger := logging.FromContext(ctx)
```
Comment (sethvargo) [inline pkg/controller/admin/admin.go:40]:
```suggestion
	logger := logging.FromContext(ctx).Named("admin")
```

File: pkg/database/user.go [hunk 10]
```diff
@@ -35,6 +35,11 @@ type User struct {
 	AdminRealms     []*Realm `gorm:"many2many:admin_realms"`
 }
 
+// CacheKey returns the key for this user in the distributed cache.
+func (u *User) CacheKey() string {
+	return fmt.Sprintf("users:by_email:%s", u.Email)
```
Comment (sethvargo) [inline pkg/database/user.go:40]: hmm... we cache users by id and email (I think)

File: cmd/server/assets/admin/new.html (hunk 2 above)
Comment (mikehelmick) [reply cmd/server/assets/admin/new.html:36]: the isn't help text for name, but I added the region code one.

File: cmd/server/assets/admin/new.html (hunk 5 above)
Comment (mikehelmick) [reply cmd/server/assets/admin/new.html:63]: added some - hard help text to write.

File: pkg/database/user.go (hunk 10 above)
Comment (mikehelmick) [reply pkg/database/user.go:40]: where this matters is in middleware/auth.go - line 77ish.
there it is cached by email

it's also interesting that the callback doesn't hit this

Comment (sethvargo) [reply pkg/database/user.go:40]: Oh hmmm... [it does](https://github.com/google/exposure-notifications-verification-server/blob/main/pkg/database/database.go#L195)? Were you seeing something else? Which user query were you running?

Comment (mikehelmick) [reply pkg/database/user.go]: the callbacks work - don't know what I Was doing wrong before

File: pkg/controller/admin/create_realm.go [hunk 11]
```diff
@@ -101,6 +104,7 @@ func (c *Controller) HandleCreateRealm() http.Handler {
 
 func (c *Controller) renderNew(ctx context.Context, w http.ResponseWriter, realm *database.Realm) {
 	m := controller.TemplateMapFromContext(ctx)
+	fmt.Printf("errors %+v", realm.Errors())
```
Comment (sethvargo) [inline pkg/controller/admin/create_realm.go:107]: Debug

**Pull request merged**